api/src/task/notifications/base_notification.py

```diff
@@ -0,0 +1,89 @@
+import logging
+from abc import abstractmethod
+from typing import Any
+from uuid import UUID
+
+import botocore.client
+from sqlalchemy import select
+
+from src.adapters.aws.pinpoint_adapter import send_pinpoint_email_raw
+from src.db.models.user_models import User, UserNotificationLog
+from src.task.notifications.constants import EmailData
+from src.task.task import Task
+
+logger = logging.getLogger(__name__)
+
+
+class BaseNotification(Task):
+
+    @abstractmethod
+    def collect_notifications(
+        self,
+    ) -> dict[UUID, list[Any]]:
```

Comment:
Two ideas to avoid `Any`
* Don't pass the values around, just call the functions, each function has to grab the values at the beginning from something it setup in `__init__`
* Generics (may or may not be better in this case, but useful to be aware of)

If you want to have the typing be a tad better here, we could use [Generics](https://typing.python.org/en/latest/reference/generics.html).

Effectively, we define this class as `class BaseNotification(Task, Generic[T])` and the derived classes are like `ClosingDateNotification(BaseNotification[UserSavedOpportunity])`. You can see a really simple version of this with our Lookup classes: https://github.com/HHS/simpler-grants-gov/blob/main/api/src/db/models/lookup/lookup.py